### DISCUSSION TITLE
Use states in functions

### DISCUSSION THREAD

**Discussion by grunzip** [dbuezas/lovelace-plotly-graph-card#407] [Q&A]
Hallo everyone,

plotly is so cool, great job.

I have a question that I haven't found yet (perhaps I have not understood the solutions): 
I would like to show the forecast of the energy stored in the battery. The data for this is available in sensors, which I can access via jinja as follows (example for 10 o'clock):
{{ state_attr('sensor.solcast_pv_forecast_prognose_today', 'detailedHourly')[10]['pv_estimate'] }}
or
{{ states['sensor.solcast_pv_forecast_today'].attributes['detailedHourly'][10]['pv_estimate'] }}

Of course, I can display this directly in plotly:

`  

    x:
      - 0
      - 1
      - 2
    'y':
      - >-
        $ex
        hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedHourly'][10]['pv_estimate']
      - >-
        $ex
        hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedHourly'][11]['pv_estimate']
      - >-
        $ex
        hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedHourly'][12]['pv_estimate']
`

But I would like to make additional calculations (battery status + forecast - estimated consumption) and wonder how I can access the data in a function. For example: How can I store the above value with the index i in y[i]:

`

    fn: |
      $fn ({ vars, state }) => {
        const x = []; 
        const y = []; 
        for (let i = 0; i <= 24; i++) {
          x[i] = i;
          y[i] = i;
        };
        vars.x = x;
        vars.y = y;
      };
    x: $fn ({ vars }) => vars.x
    'y': $fn ({ vars }) => vars.y
`

Tank you very much
Roland

**Comment by dbuezas** [comment]
Cool plot-project!
You can access it like this:
```yaml
x: |
  $fn ({ vars, hass }) => {
    const some_other_calculation = xxxxxx;
    return some_other_calculation + vars.y + hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedHourly'][10]['pv_estimate']
  }
```

Or this is the same:

```yaml
x: |
  $ex { 
    const some_other_calculation = xxxxxx;
    return some_other_calculation + vars.y + hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedHourly'][10]['pv_estimate']
  }
```

$ex is exactly the same as $fn, except that the parameters are added automatically.

The third (maybe best) solution is:

```yaml
entity: sensor.solcast_pv_forecast_prognose_heute
y: |
  $ex {
     const est = meta.detailedHourly[10].pv_estimate;
     ....
     return whatever;
  }
```

the `meta` attribute holds `hass.states[entity_id].attributes`, which may be shorter to use

**Reply by grunzip** [reply]
Thank you very much!
I didn't know, that 'hass' has to be a parameter. Is it also possible to substitute '10' by a variable?

**Comment by grunzip** [comment]
My results so far
![Screenshot 2024-04-14 151503](https://github.com/dbuezas/lovelace-plotly-graph-card/assets/148907343/bc174a92-3de6-44d3-b562-431b361d7b9f)


```
- entity: null
    name: Akku
    fn: |
      $fn ({ vars, hass }) => {
        const x = []; 
        const y = []; 
        const z = [];
        energyStored = parseFloat(hass.states['sensor.foxess_residual_energy'].state);
        batteryCapacity = 10.96;
        var jetzt = new Date(),
        h = 2*jetzt.getHours();
        m = 2*jetzt.getMinutes();
        s = 2*jetzt.getSeconds();
        h_m = h + m/60;
        h = Math.floor(h_m);
        load = parseFloat(hass.states['input_number.prognose_pauschaler_stromverbrauch'].state) / 1000;
        for (let i = 0; i <= 47; i++) {
          z[i] = hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+48] = hass.states['sensor.solcast_pv_forecast_prognose_morgen'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+96] = hass.states['sensor.solcast_pv_forecast_prognose_tag_3'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+144] = hass.states['sensor.solcast_pv_forecast_prognose_tag_4'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+192] = hass.states['sensor.solcast_pv_forecast_prognose_tag_5'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+240] = hass.states['sensor.solcast_pv_forecast_prognose_tag_6'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+288] = hass.states['sensor.solcast_pv_forecast_prognose_tag_7'].attributes['detailedForecast'][i]['pv_estimate'];
        }
        y[h] = energyStored;
        x[h] = h_m/2;
        for (let i = h+1; i <= 336; i++) {
          x[i] = (i)/2;
          if (i == h+1) 
            {y[i] = y[i-1] + (z[i-1] - load) * (i - h_m)/2} 
          else 
            {y[i] = y[i-1] + (z[i-1] - load) / 2};
          y[i] = Math.min(y[i], batteryCapacity);
        };
        vars.x = x;
        vars.y = y;
      };
    x: $fn ({ vars }) => vars.x
    'y': $fn ({ vars }) => vars.y
    line:
      color: cornflowerblue
  - entity: null
    name: PV
    fn: |
      $fn ({ vars, hass }) => {
        const x = []; 
        const y = []; 
        const z = [];
        energyStored = parseFloat(hass.states['sensor.foxess_residual_energy'].state);
        batteryCapacity = 10.96;
        var jetzt = new Date(),
        h = 2*jetzt.getHours();
        m = 2*jetzt.getMinutes();
        s = 2*jetzt.getSeconds();
        h_m = h + m/60;
        h = Math.floor(h_m);
        load = parseFloat(hass.states['input_number.prognose_pauschaler_stromverbrauch'].state) / 1000;
        for (let i = 0; i <= 47; i++) {
          z[i] = hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+48] = hass.states['sensor.solcast_pv_forecast_prognose_morgen'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+96] = hass.states['sensor.solcast_pv_forecast_prognose_tag_3'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+144] = hass.states['sensor.solcast_pv_forecast_prognose_tag_4'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+192] = hass.states['sensor.solcast_pv_forecast_prognose_tag_5'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+240] = hass.states['sensor.solcast_pv_forecast_prognose_tag_6'].attributes['detailedForecast'][i]['pv_estimate'];
          z[i+288] = hass.states['sensor.solcast_pv_forecast_prognose_tag_7'].attributes['detailedForecast'][i]['pv_estimate'];
        }
        y[h] = energyStored;
        x[h] = h_m/2;
        for (let i = h+1; i <= 336; i++) {
          x[i] = (i)/2;
          if (i == h+1) 
            {y[i] = y[i-1] + (z[i-1] - load) * (i - h_m)/2} 
          else 
            {y[i] = y[i-1] + (z[i-1] - load) / 2};
        };
        vars.x = x;
        vars.y = y;
      };
    x: $fn ({ vars }) => vars.x
    'y': $fn ({ vars }) => vars.y
    line:
      dash: dot
      color: cornflowerblue``

**Reply by dbuezas** [reply]
You can shorten it with :
z = hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes.detailedHourly.map(d=>d.pv_estimate)
```

**Reply by grunzip** [reply]
Thank you so much for your response, but your version leads to

``Error: at [entities.0.fn]: (d >= d.pv_estimate) is not a function``

Not sure, if this is the problem, but 
``z[1] = hass.states['sensor.solcast_pv_forecast_prognose_heute'].attributes.detailedHourly[1].pv_estimate;``
results not in an error. But trials to replace ``[1]`` with ``.1`` leads to an error like
``Error: at [entities.0.fn]: unexpected token: numeric literal``

**Reply by dbuezas** [reply]
It's `=>` not `>=`

**Reply by dbuezas** [reply]
It means function, what you wrote means greater or equal

**Reply by grunzip** [reply]
Sorry - my mistake, I tried several versions. With your version the message is

``Error: at [entities.0.fn]: invalid assignment to const 'z'``

I have updated the version that works fine for me.
[https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/407#discussioncomment-9087781](url)

**Reply by dbuezas** [reply]
Yes, you already defined z above with const. But your updated code with for loops looks good
